src/log_batch.rs

```diff
@@ -932,6 +941,114 @@ fn verify_checksum_with_signature(buf: &[u8], signature: Option<u32>) -> Result<
+            ..
+        }) = &item.content
+        {
+            if *op_type == OpType::Put
+                && value.as_ref().unwrap().len() == ATOMIC_GROUP_VALUE_LEN
+                && *key == crate::make_internal_key(ATOMIC_GROUP_KEY)
+            {
+                let value = &mut value.as_ref().unwrap().as_slice();
+                let id = codec::decode_u64(value).unwrap();
+                return Some((id, AtomicGroupStatus::from_u8(value[0]).unwrap()));
+            }
+        }
+        None
+    }
+}
+
+/// Group multiple log batches as an atomic operation.
+///
+/// Caveats:
+/// (1) The atomicity is provided at persistent level. This means, once an
+/// atomic group fails, the in-memory value will be inconsistent with what can
+/// be recovered from on-disk data files.
+/// (2) The recovery replay order will be different from original write order.
+/// Log batches in a completed atomic group will be replayed as if they were
+/// written together at an arbitary time point within the group.
+/// (3) Atomic group is implemented by embedding normal key-values into user
+/// writes. These keys have internal key prefix and will not be replayed into
+/// memtable. However, when read by an older version, they will behave as user
+/// keys. They may also belong to Raft Group that doesn't exist before.
+///
+/// In practice, we only use atomic group for rewrite operation. (In fact,
+/// atomic group markers in append queue are simply ignored.) Rewrite doesn't
+/// change the value of entries, just locations. So first issue doesn't affect
+/// correctness. There could only be one worker doing the rewrite. So second
+/// issue doesn't change observed write order because there's no mixed write.
+pub(crate) struct AtomicGroupBuilder {
+    id: u64,
+    begin: bool,
+}
+
+impl Default for AtomicGroupBuilder {
+    fn default() -> Self {
+        Self {
+            // We only care there's no collision between concurrent groups.
+            id: ATOMIC_GROUP_ID.fetch_add(1, Ordering::Relaxed),
+            begin: false,
+        }
+    }
+}
+
+impl AtomicGroupBuilder {
+    /// Each log batch can only carry one atomic group marker. If multiple are
+    /// present only the first is recognized.
+    pub fn begin(&mut self, lb: &mut LogBatch) {
+        fail::fail_point!("atomic_group::begin");
+        assert!(!self.begin);
+        let mut s = Vec::with_capacity(ATOMIC_GROUP_VALUE_LEN);
+        s.encode_u64(self.id).unwrap();
+        s.push(AtomicGroupStatus::Begin as u8);
+        lb.put(self.id, crate::make_internal_key(ATOMIC_GROUP_KEY), s);
+        self.begin = true;
+    }
+
+    pub fn add(&self, lb: &mut LogBatch) {
+        fail::fail_point!("atomic_group::add");
+        assert!(self.begin);
+        let mut s = Vec::with_capacity(ATOMIC_GROUP_VALUE_LEN);
+        s.encode_u64(self.id).unwrap();
+        s.push(AtomicGroupStatus::Middle as u8);
+        lb.put(self.id, crate::make_internal_key(ATOMIC_GROUP_KEY), s);
```

Comment:
Since the id is already a part of key, why encode it in value again?

Reply:
Good catch.